perses/dispersed/utils.py

```diff
@@ -556,6 +556,24 @@ def create_endstates_from_real_systems(htf, for_testing=False):
         for index in sorted(indices_to_remove, reverse=True):
             hybrid_system.removeForce(index)
 
+            # Set defaults for global parameters depending on the factory
+            htf_class = htf.__class__.__name__
+            for force_index, force in enumerate(list(hybrid_system.getForces())):
+                if hasattr(force, 'getNumGlobalParameters'): # Only custom forces will have global parameters to set
+                    for parameter_index in range(force.getNumGlobalParameters()):
+                        global_parameter_name = force.getGlobalParameterName(parameter_index)
+                        if global_parameter_name[0:7] == 'lambda_':
+                            if htf_class == 'HybridTopologyFactory':
+                                force.setGlobalParameterDefaultValue(parameter_index, lambda_val)
+                            elif htf_class == 'RESTCapableHybridTopologyFactory':
+                                if 'old' in global_parameter_name:
+                                    force.setGlobalParameterDefaultValue(parameter_index, 1 - lambda_val)
+                                elif 'new' in global_parameter_name:
+                                    force.setGlobalParameterDefaultValue(parameter_index, lambda_val)
```

Comment:
Just wondering if this would be something that the actual HTF objects should do on their side, rather than having all the logic and conditionals here. Or is this something that's only accessible or only makes sense when creating the endstates? (Non-blocking)

Reply:
Yes, `setGlobalParameterDefaultValue()` can be called on the HTF side. 

Let me walk through what the code above does and try to write a helper function that would simplify the code. We are creating two hybrid systems: one for the lambda = 0 endstate and one for the lambda = 1 endstate. Therefore, the hybrid systems should have different default values (for lambda = 0, the default values will be 0 and for lambda = 1, the default values will be 1). However, the challenge is that the global parameters for the different factories work differently. 
- For the `HybridTopologyFactory`, we want `lambda_*` to be 0 for lambda = 0 and 1 for lambda = 1. 
- However, for `RESTCapableHybridTopologyFactory`, we decided that:
    - for lambda = 0,  `lambda_*_old` should be 1 and `lambda_*_new` should be 0. 
    - for lambda = 1, `lambda_*_old` should be 0 and `lambda_*_new` should be 1.

Here is are suggested helper functions that we could add to the factories to simplify the logic above.
```python
# Helper function for HybridTopologyFactory
def set_lambdas_for_endstate(self, endstate):
    for force_index, force in enumerate(list(hybrid_system.getForces())):
        if hasattr(force, 'getNumGlobalParameters'):
            for parameter_index in range(force.getNumGlobalParameters()):
                global_parameter_name = force.getGlobalParameterName(parameter_index)
                if global_parameter_name[0:7] == 'lambda_':
                    force.setGlobalParameterDefaultValue(parameter_index, endstate)

# Helper function for RESTCapableHybridTopologyFactory
def set_lambdas_for_endstate(self, endstate):
    for force_index, force in enumerate(list(hybrid_system.getForces())):
        if hasattr(force, 'getNumGlobalParameters'):
            for parameter_index in range(force.getNumGlobalParameters()):
                global_parameter_name = force.getGlobalParameterName(parameter_index)
                if global_parameter_name[0:7] == 'lambda_':
                    if 'old' in global_parameter_name:
                        force.setGlobalParameterDefaultValue(parameter_index, 1 - endstate)
                    elif 'new' in global_parameter_name:
                        force.setGlobalParameterDefaultValue(parameter_index, endstate)
```
which would simplify the above code to:
```python
# Set defaults for global parameters depending on the factory
htf_class = htf.__class__.__name__
htf.set_lambdas_for_endstate(lambda_val)
```

Reply:
I have been thinking about this and I think the best approach is probably just to leave it like this and refactor to a more factory-like pattern for the next 0.11.0 release, hopefully without breaking any API compatibility. I can already see many things that can be improved in the whole structure of the classes and methods, but I don't want to make such big changes right now. This is good enough for the current state.